src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperResolutionResultJsonConverter.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Razor.Language;
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Razor.Serialization
+{
+    internal class TagHelperResolutionResultJsonConverter : JsonConverter
+    {
+        public static readonly TagHelperResolutionResultJsonConverter Instance = new TagHelperResolutionResultJsonConverter();
+
+        public override bool CanConvert(Type objectType)
+        {
+            return typeof(TagHelperResolutionResult).IsAssignableFrom(objectType);
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader.TokenType != JsonToken.StartObject || !reader.Read())
+            {
+                return null;
+            }
+
+            var descriptors = new List<TagHelperDescriptor>();
+            var descriptorConverter = TagHelperDescriptorJsonConverter.Instance;
+
+            if (reader.TokenType != JsonToken.PropertyName ||
```

Comment:
This whole if statement is quite confusing. Could you break it into many with comments